src/tests/unit_tls.cpp

```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
+                           input.resize(needed);
+                           Test::rng().randomize(input.data(), input.size());
+                           client->received_data(input.data(), input.size());
+                           }
+                        }
+                     catch(std::exception&)
+                        {
+                        result.test_note("corruption caused client exception");
+                        }
                      }
-                  catch(std::exception& e)
+                  else
                      {
-                     result.test_failure("client error", e.what());
+                     try
+                        {
+                        size_t needed = client->received_data(input.data(), input.size());
+                        result.test_eq("full packet received", needed, 0);
+                        }
+                     catch(std::exception& e)
+                        {
+                        result.test_failure("client error", e.what());
+                        }
                      }
-                  }
 
-               continue;
-               }
+                  continue;
+                  }
 
-            // If we corrupted a DTLS application message, resend it:
-            if(client.is_active() && corrupt_client_data && server_recv.empty())
-               client.send(client_sent);
-            if(server.is_active() && corrupt_server_data && client_recv.empty())
-               server.send(server_sent);
+               // If we corrupted a DTLS application message, resend it:
+               if(client->is_active() && corrupt_client_data && server_recv.empty())
+                  client->send(client_sent);
+               if(server->is_active() && corrupt_server_data && client_recv.empty())
+                  server->send(server_sent);
 
-            if(client_recv.size())
-               {
-               result.test_eq("client recv", client_recv, server_sent);
-               }
+               if(client_recv.size())
+                  {
+                  result.test_eq("client recv", client_recv, server_sent);
+                  }
 
-            if(server_recv.size())
-               {
-               result.test_eq("server recv", server_recv, client_sent);
-               }
+               if(server_recv.size())
+                  {
+                  result.test_eq("server recv", server_recv, client_sent);
+                  }
 
-            if(client.is_closed() && server.is_closed())
-               break;
+               if(client->is_closed() && server->is_closed())
+                  break;
 
-            if(server_recv.size() && client_recv.size())
-               {
-               Botan::SymmetricKey client_key = client.key_material_export("label", "context", 32);
-               Botan::SymmetricKey server_key = server.key_material_export("label", "context", 32);
+               if(server_recv.size() && client_recv.size())
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)